File: donuts/donut.cs
                using static System
            .Math;int k;double A=0,B=0,
         i,j; double[] z=new double[1760];
       char[] b=new char[1760];Console.Clear
     ();for(;;){for(k=0;k<1760;k++){ z[k]=0; b
    [k]=' ';}for(j=0;PI*2>j;j+=0.07){for(i=0;PI
  *2>i;i+=0.02){ double sini=Sin(i), cosj=Cos(j),
  sinA=Sin(A),sinj=Sin(j),cosA=Cos(A),cosj2=cosj+
 2, mess =1 / (sini*           cosj2*sinA + sinj *
cosA+5),cosi=Cos(i               ),cosB=Cos(B),sinB
=Sin(B), t=sini*                   cosj2*cosA-sinj*
sinA;int x=(int)                   (40 + 30*mess* (
cosi*cosj2*cosB-                   t*sinB)),y=(int)
(12 + 15*mess* (                   cosi*cosj2*sinB+
t*cosB)),o=x+80*                   y, N=(int)(8* ((
sinj*sinA - sini *               cosj*cosA) *cosB -
 sini*cosj*sinA-sinj           *cosA-cosi*cosj*Sin
  (B)));if(22>y&&y>0&&x>0&&80>x&&mess>z[o]){z[o]=
  mess; b[o]=".,-~:;=!*#$@"[N>0?N:0]; }}}Console.
    Write("\x1b[d");for(k=0;1761>k;k++)Console.
     Write(k%80!=0?b[k]:'\n');A+=0.04;B+=0.02;
       }/*..................................
         ........... Made  for ...........
            ........ .NET  7.0 ........
                .................*/

File: donuts/donut.ml
                   let a,b, s,u,
             tS,pS,r,v,m= ref 0.,ref
          0.,40., 40,0.03,0.02,1.,2.,5.
       in let k=s *.m*.3. /.(8.*.(r+.v))in
     let rf a b=let g,h,n,q,o,zB,t=cos!a,sin
    !a,cos!b, sin!b,Array.make_matrix u u ' '
   ,Array.make_matrix u u 0. ,ref 0. in while(
  !t< 6.28)do let c,d,p = cos!t,sin!t,ref 0. in
 while(!p<6.28)do let e,f,i,j=cos!p,sin!p,v+.r*.
c,r*.d in let x,y,z           =i*.(n*.e+.h*.q*.f)
-.j*.g*.q,i*.(q*.e             -.h*.n*.f)+.j*.g*.
n,m+.g*.i*.f+.j*.               h in let w=1. /.z
in let xp, yp, l=               truncate(s/.2.+.k
*.w*.x),truncate(               s/.2.-.k*.w*.y),e
*.c*.q-.g*.c*.f-.h             *.d+.n*.(g*.d-.c*.
h*.f)in if(0.<l)&&(           zB.(xp).(yp)<w)then
 (let lI=truncate(l*.8.)in zB.(xp).(yp)<-w;o.(xp
  ).(yp)<-".,-~:;=!*#$@".[lI]); p:=!p+.pS; done
   ;t:= !t +. tS; done; for j =0 to u-1 do for
    i=0 to u-1 do print_char o.(i).(j); done;
     print_endline""done;in while true do rf
       a b;print_string"\x1b[2J\x1b[H";a:=
          !a+. tS; b:= !b +. pS; done;;
             (*-------Made by-------)
                  (sully_vian*)